radicle-seed/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+        let mode = self.config.mode;
+
+        // Spawn the background peer thread.
+        tokio::spawn(future);
+
+        // Track already-known URNs.
+        Node::initialize_tracker(&mode, &api).await?;
+
+        // Listen on gossip events. As soon as a peer announces something of interest,
+        // we check if we should track it.
+        while let Some(event) = events.next().await {
+            match event {
+                ProtocolEvent::Gossip(gossip::Info::Has(gossip::Has { provider, val })) => {
+                    let urn = &val.urn;
+                    let peer_id = &provider.peer_id;
+
+                    tracing::info!("Discovered new URN {} from peer {}", urn, peer_id);
+
+                    if mode.is_trackable(peer_id, urn) {
+                        // Attempt to track, but keep going if it fails.
+                        Node::track_project(&api, urn, &provider).ok();
+                    }
+                },
+                _ => {},
+            }
+        }
+        Ok(())
+    }
+
+    /// Attempt to track a project.
+    fn track_project(
+        api: &PeerApi<Signer>,
+        urn: &RadUrn,
+        peer_info: &PeerInfo<std::net::IpAddr>,
+    ) -> Result<(), Error> {
+        let peer_id = &peer_info.peer_id;
+        let url = urn.clone().into_rad_url(peer_id.clone());
+        let port = peer_info.advertised_info.listen_port;
+        let addr_hints = peer_info
+            .seen_addrs
+            .iter()
+            .map(|a: &std::net::IpAddr| (*a, port).into());
+
+        let result = api
+            .storage()
+            .clone_repo::<project::ProjectInfo, _>(url, addr_hints)
+            .and_then(|_| api.storage().track(urn, &peer_id));
+
+        match &result {
+            Ok(()) => {
+                tracing::info!("Successfully tracked project {} from peer {}", urn, peer_id,);
+            },
+            Err(err) => {
+                tracing::debug!(
+                    "Error tracking project {} from peer {}: {}",
+                    urn,
+                    peer_id,
+                    err
+                );
+            },
+        }
+        result.map_err(Error::from)
+    }
+
+    /// Attempt to track initial URN list, if any.
+    async fn initialize_tracker(mode: &Mode, api: &PeerApi<Signer>) -> Result<(), Error> {
+        // Start by tracking specified projects if we need to.
+        match &mode {
+            Mode::TrackUrns(urns) => {
+                tracing::info!("Initializing tracker with {} URNs..", urns.len());
+
+                for urn in urns {
+                    let mut peers = api.providers(urn.clone()).await;
+                    // Attempt to track until we succeed.
+                    while let Some(peer) = peers.next().await {
+                        if Node::track_project(&api, urn, &peer).is_ok() {
+                            break;
+                        }
+                    }
+                }
```

Comment:
```suggestion
                futures::stream::iter(urns)
                    .for_each_concurrently(Some(urns.len()), |urn| {
                        let mut peers = api.providers(urn.clone()).await;
                        tokio::time::timeout(Duration::from_secs(120), {
                            // Attempt to track until we succeed.
                            while let Some(peer) = peers.next().await {
                                if Node::track_project(&api, urn, &peer).is_ok() {
                                    break;
                                }
                            }
                        }).unwrap_or_else(|| tracing::warn!("Timeout trying to track {}", urn))
                    }) 
```

(not sure if typechecks)